sdk/src/azure/core/az_http_policy_retry.c

```diff
@@ -84,67 +75,77 @@ AZ_INLINE AZ_NODISCARD int32_t _az_uint32_span_to_int32(az_span span)
   return value < INT32_MAX ? (int32_t)value : INT32_MAX;
 }
 
+AZ_INLINE AZ_NODISCARD bool _az_http_policy_retry_should_retry_http_response_code(
+    az_http_status_code http_response_code)
+{
+  switch (http_response_code)
+  {
+    case AZ_HTTP_STATUS_CODE_REQUEST_TIMEOUT:
+    case AZ_HTTP_STATUS_CODE_TOO_MANY_REQUESTS:
+    case AZ_HTTP_STATUS_CODE_INTERNAL_SERVER_ERROR:
+    case AZ_HTTP_STATUS_CODE_BAD_GATEWAY:
+    case AZ_HTTP_STATUS_CODE_SERVICE_UNAVAILABLE:
+    case AZ_HTTP_STATUS_CODE_GATEWAY_TIMEOUT:
+      return true;
+    default:
+      return false;
+  }
+}
+
 AZ_INLINE AZ_NODISCARD az_result _az_http_policy_retry_get_retry_after(
     az_http_response* ref_response,
-    az_http_status_code const* status_codes,
     bool* should_retry,
     int32_t* retry_after_msec)
 {
   az_http_response_status_line status_line = { 0 };
   _az_RETURN_IF_FAILED(az_http_response_get_status_line(ref_response, &status_line));
-  az_http_status_code const response_code = status_line.status_code;
 
-  for (; *status_codes != _az_HTTP_STATUS_CODE_END_OF_LIST; ++status_codes)
+  if (!_az_http_policy_retry_should_retry_http_response_code(status_line.status_code))
   {
-    if (*status_codes != response_code)
-    {
-      continue;
-    }
+    *should_retry = false;
+    *retry_after_msec = -1;
+    return AZ_OK;
```

Comment:
Are you sure we should be returning `AZ_OK` in this case rather than bubbling up some error `az_result` to the caller?

Reply:
Yes, in this place, we only return failures if we failed to parse the response status line.
This is a local function. If the function that calls it has exhausted all the retry attempts - then the error code gets returned to the customer.

Reply:
> If the function that calls it has exhausted all the retry attempts - only then we fail with error.

Is that what's happening? I looks like if `should_retry` is false, we just return `AZ_OK` here:
https://github.com/Azure/azure-sdk-for-c/blob/dbb8393c63c7c7278cd30d86e3a978b797753fd6/sdk/src/azure/core/az_http_policy_retry.c#L194-L197

Reply:
And that is fine too - we return the latest response. We fail with != AZ_OK only when there is a critical error - we can't parse the HTTP response, we can't mark the retry headers, we can't initialize the response, we can't remove the retry headers.
If after 5 retry attempts the response you get from the server is still with `AZ_HTTP_STATUS_CODE_BAD_GATEWAY` code, that response ultimately gets returned to the client.

Reply:
.NET, for example, also behaves this way: https://github.com/Azure/azure-sdk-for-net/blob/d879da95f4c642031089ea2208f25de55231a98a/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs#L110